engines/ceg_prophecy_arrays.cpp

```diff
@@ -0,0 +1,459 @@
+  initialize();
+}
+
+ProverResult CegProphecyArrays::prove()
+{
+  ProverResult res = ProverResult::FALSE;
+  while (res == ProverResult::FALSE) {
+    // Refine the system
+    // heuristic -- stop refining when no new axioms are needed.
+    do {
+      if (!refine()) {
+        return ProverResult::FALSE;
+      }
+      reached_k_++;
+    } while (num_added_axioms_);
+
+    Property latest_prop(abs_ts_, solver_->make_term(Not, bad_));
+    shared_ptr<Prover> prover =
+        make_prover(e_, latest_prop, solver_->get_solver_enum(), options_);
+    res = prover->prove();
+  }
+
+  return res;
+}
+
+ProverResult CegProphecyArrays::check_until(int k)
+{
+  ProverResult res = ProverResult::FALSE;
+  while (res == ProverResult::FALSE && reached_k_ <= k) {
+    // Refine the system
+    // heuristic -- stop refining when no new axioms are needed.
+    do {
+      if (!refine()) {
+        return ProverResult::FALSE;
+      }
+      reached_k_++;
+    } while (num_added_axioms_ && reached_k_ <= k);
+
+    Property latest_prop(abs_ts_, solver_->make_term(Not, bad_));
+    shared_ptr<Prover> prover =
+        make_prover(e_, latest_prop, solver_->get_solver_enum(), options_);
+
+    res = prover->check_until(k);
+  }
+
+  if (res == ProverResult::FALSE) {
+    // can't count on false result over abstraction when only checking up until
+    // a bound
+    return ProverResult::UNKNOWN;
+  }
+
+  return res;
+}
+
+void CegProphecyArrays::initialize()
+{
+  super::initialize();
+  abstract();
+
+  bool contains_arrays = false;
+  for (auto sv : conc_ts_.statevars()) {
+    if (sv->get_sort()->get_sort_kind() == ARRAY) {
+      contains_arrays = true;
+      break;
+    }
+  }
+
+  for (auto iv : conc_ts_.inputvars()) {
+    if (iv->get_sort()->get_sort_kind() == ARRAY) {
+      contains_arrays = true;
+      break;
+    }
+  }
+
+  if (!contains_arrays) {
+    throw PonoException("Ran CegProphecyArrays on system without arrays.");
+  }
+}
+
+void CegProphecyArrays::abstract()
+{
+  // the ArrayAbstractor already abstracted the transition system on
+  // construction -- only need to abstract bad
+  bad_ = aa_.abstract(bad_);
+
+  // the abstract system should have all the same state and input variables
+  // but abstracted
+  // plus it will have some new variables for the witnesses and lambdas
+  assert(abs_ts_.statevars().size() >= conc_ts_.statevars().size());
+  assert(abs_ts_.inputvars().size() >= conc_ts_.inputvars().size());
+}
+
+bool CegProphecyArrays::refine()
+{
+  num_added_axioms_ = 0;
+  // TODO use ArrayAxiomEnumerator and modifiers to refine the system
+  // create BMC formula
+  Term abs_bmc_formula = get_bmc_formula(reached_k_ + 1);
+
+  // check array axioms over the abstract system
+  if (!aae_.enumerate_axioms(abs_bmc_formula, reached_k_ + 1)) {
+    // concrete CEX
+    return false;
+  }
+
+  UnorderedTermSet consecutive_axioms = aae_.get_consecutive_axioms();
+  AxiomVec nonconsecutive_axioms = aae_.get_nonconsecutive_axioms();
```

Comment:
```suggestion
  const AxiomVec &nonconsecutive_axioms = aae_.get_nonconsecutive_axioms();
```

Reply:
This is the same as the one above. It's copied so that it can be modified in place when being reduced. I could change the interface of the reducers, although then we'd have to have two versions -- the reduced one and the unreduced one. As it is now, it just gets edited when reduced. Thoughts?